go/test/endtoend/vreplication/fk_ext_test.go

```diff
@@ -0,0 +1,444 @@
+* Sets up a source keyspace, to simulate the external database, with a parent and child table with a foreign key constraint.
+* Creates a target keyspace with two shards, the Vitess keyspace, into which the source data is imported.
+* Imports the data using MoveTables. This uses the atomic copy flow
+to test that we can import data with foreign key constraints and that data is kept consistent even after the copy phase
+since the tables continue to have the FK Constraints.
+* Creates a new keyspace with two shards, the Vitess keyspace, into which the data is migrated using MoveTables.
+* Materializes the parent and child tables into a different keyspace.
+* Reshards the keyspace from 2 shards to 3 shards.
+* Reshards the keyspace from 3 shards to 1 shard.
+
+We drop constraints from the tables from some replicas to simulate a replica that is not doing cascades in
+innodb, to confirm that vtgate's fkmanaged mode is working properly.
+*/
+
+func TestFKExt(t *testing.T) {
+	setSidecarDBName("_vt")
+
+	// Ensure that there are multiple copy phase cycles per table.
+	extraVTTabletArgs = append(extraVTTabletArgs, "--vstream_packet_size=256", "--queryserver-config-schema-change-signal")
+	extraVTGateArgs = append(extraVTGateArgs, "--schema_change_signal=true", "--planner-version", "Gen4")
+	defer func() { extraVTTabletArgs = nil }()
+	initFKExtConfig(t)
+
+	cellName := fkextConfig.cell
+	cells := []string{cellName}
+	vc = NewVitessCluster(t, t.Name(), cells, fkextConfig.ClusterConfig)
+
+	require.NotNil(t, vc)
+	allCellNames = cellName
+	defaultCellName := cellName
+	defaultCell = vc.Cells[defaultCellName]
+	cell := vc.Cells[cellName]
+
+	defer vc.TearDown(t)
+
+	sourceKeyspace := fkextConfig.sourceKeyspaceName
+	vc.AddKeyspace(t, []*Cell{cell}, sourceKeyspace, "0", FKExtSourceVSchema, FKExtSourceSchema, 0, 0, 100, nil)
+
+	vtgate = cell.Vtgates[0]
+	require.NotNil(t, vtgate)
+	err := cluster.WaitForHealthyShard(vc.VtctldClient, sourceKeyspace, "0")
+	require.NoError(t, err)
+	require.NoError(t, vtgate.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.primary", sourceKeyspace, "0"), 1, shardStatusWaitTimeout))
+
+	vtgateConn = getConnection(t, vc.ClusterConfig.hostname, vc.ClusterConfig.vtgateMySQLPort)
+	defer vtgateConn.Close()
+	verifyClusterHealth(t, vc)
+
+	lg = &SimpleLoadGenerator{}
+	lg.Init(context.Background(), vc)
+	lg.SetDBStrategy("vtgate", fkextConfig.sourceKeyspaceName)
+	if lg.Load() != nil {
+		t.Fatal("Load failed")
+	}
+	if lg.Start() != nil {
+		t.Fatal("Start failed")
+	}
+	t.Run("Import from external db", func(t *testing.T) {
+		// Import data into vitess from sourceKeyspace to target1Keyspace, both unsharded.
+		importIntoVitess(t)
+	})
+
+	t.Run("MoveTables from unsharded to sharded keyspace", func(t *testing.T) {
+		// Migrate data from target1Keyspace to the sharded target2Keyspace. Streams only from primaries
+		// for one shard and replica for the other shard. Constraints have been dropped from this replica.
```

Comment:
I'm not sure what "Streams only from primaries for one shard and replica for the other shard" mean here, as we're importing from an unsharded keyspace?

Reply:
Comment was incorrect/ambiguous. They were in the context of the following materialize workflow which is from the sharded `target2` to the unsharded `target1`. I updated the comments to reflect that.